display.py

```diff
@@ -21,271 +24,309 @@ def font(family='', bold=False, italic=False):
+        self.ctx.rectangle(x - size/2, y - size/2,
                            size, size)
         self.ctx.fill()
 
-    def draw_circle(self, size=10, *a, **kw):
-        _draw_circle(self.ctx, self.x0, self.y0, size, *a, **kw)
+    def draw_circle(self, radius=(1/3), coords=None, **kw):
+        center = self.display.convert_coords(coords)
+        draw_circle(self.ctx, center, radius, **kw)
 
 class EdgeContext(object):
-    def __init__(self, surface, edge, model, scale):
-        self.surface = surface
+    def __init__(self, display, ctx, edge, model, extra):
+        self.display = display
+        self.ctx = ctx
         self.edge = edge
         self.model = model
-        self.scale = scale
-        self.ctx = cairo.Context(surface)
+        self.extra = extra
 
     @property
     def val(self):
         return str(self.model[self.edge.var])
 
-    @property
-    def x0(self):
-        return transform_x(self.gx, self.scale)
-
-    @property
-    def y0(self):
-        return transform_y(self.gy, self.scale)
-
-    @property
-    def gx(self):
-        return self.edge.x
-
-    @property
-    def gy(self):
-        return self.edge.y
-
     def draw(self, width=1, color=(0, 0, 0, 1)):
-        self.ctx.set_line_width(width)
         self.ctx.set_line_cap(cairo.LINE_CAP_SQUARE)
-        r, g, b, a = color
-        self.ctx.set_source_rgba(b, g, r, a)
-        self.ctx.move_to(self.x0, self.y0)
-        self.ctx.line_to(self.x1, self.y1)
+        # transform width into user space
+        real_width = convert_stroke_width(self.ctx, width)
+        self.ctx.set_line_width(real_width)
+        self.ctx.set_source_rgba(*color)
+        self.ctx.move_to(0,0)
+        self.ctx.line_to(1,0)
         self.ctx.stroke()
 
-    def draw_circle(self, size=None, *a, **kw):
-        # TODO: consistent name (circle vs. draw_circle)
-        if not size:
-            size = self.scale / 2.3
-        _draw_circle(self.ctx, (self.x0 + self.x1)/2, (self.y0 + self.y1)/2, size, *a, **kw)
-
-
-class HorizEdgeContext(EdgeContext):
-    def __init__(self, *a, **kw):
-        super(HorizEdgeContext, self).__init__(*a, **kw)
-
-    @property
-    def x1(self):
-        return transform_x(self.gx + 1, self.scale)
-
-    @property
-    def y1(self):
-        return self.y0
-
-class VertEdgeContext(EdgeContext):
-    def __init__(self, *a, **kw):
-        super(VertEdgeContext, self).__init__(*a, **kw)
-
-    @property
-    def x1(self):
-        return self.x0
-
-    @property
-    def y1(self):
-        return transform_y(self.gy + 1, self.scale)
+    def draw_text(self, text, coords=None, **kw):
+        coords = self.display.convert_coords(coords, default=(0.5, 0))
+        draw_text(self.ctx, text, coords, **kw)
 
 class CellContext(object):
-    def __init__(self, surface, cell, model, scale):
-        self.surface = surface
-        self.ctx = cairo.Context(surface)
+    def __init__(self, display, ctx, cell, model, extra):
```

Comment:
This `extra` business, which is threaded though a number of different places, only exists because of the multi-frame use case? Why not just stick to giving the context a frame index, the way it used to work with framed grids? This is just more complexity that doesn't make the consumer's code any simpler.

Reply:
I certainly could reduce the `extra` to a simple frame index. That seems reasonable. Honestly I'm not sure why that didn't occur to me. The only person who really knows how to interpret the `extra` is the person who created it, and they'll still be able to figure out what it was supposed to be from the frame index.